slasher/db/kv/spanner.go

```diff
@@ -0,0 +1,306 @@
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.unmarshalSpan")
+	defer span.End()
+	r := types.Span{}
+	if len(enc) != 7 {
+		return r, errors.New("wrong data length for min max span")
+	}
+	r.MinSpan = bytesutil.FromBytes2(enc[:2])
+	r.MaxSpan = bytesutil.FromBytes2(enc[2:4])
+	sigB := [2]byte{}
+	copy(sigB[:], enc[4:6])
+	r.SigBytes = sigB
+	r.HasAttested = bytesutil.ToBool(enc[6])
+	return r, nil
+}
+
+// convert the span struct into a flat array
+func marshalSpan(span types.Span) []byte {
+	return append(append(append(
+		bytesutil.Bytes2(uint64(span.MinSpan)),
+		bytesutil.Bytes2(uint64(span.MaxSpan))...),
+		span.SigBytes[:]...),
+		bytesutil.FromBool(span.HasAttested))
+}
+
+// EpochSpansMap accepts epoch and returns the corresponding spans map epoch=>spans
+// for slashing detection.
+// Returns nil if the span map for this validator index does not exist.
+func (db *Store) EpochSpansMap(ctx context.Context, epoch uint64) (map[uint64]types.Span, error) {
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.EpochSpansMap")
+	defer span.End()
+	if db.spanCacheEnabled {
+		v, ok := db.spanCache.Get(epoch)
+		spanMap := make(map[uint64]types.Span)
+		if ok {
+			spanMap, ok = v.(map[uint64]types.Span)
+			if !ok {
+				return nil, cacheTypeMismatchError(v)
+			}
+			return spanMap, nil
+		}
+	}
+	var err error
+	var spanMap map[uint64]types.Span
+	err = db.view(func(tx *bolt.Tx) error {
+		b := tx.Bucket(validatorsMinMaxSpanBucket)
+		epochBucket := b.Bucket(bytesutil.Bytes8(epoch))
+		if epochBucket == nil {
+			return nil
+		}
+		keysLength := epochBucket.Stats().KeyN
+		spanMap = make(map[uint64]types.Span, keysLength)
+		return epochBucket.ForEach(func(k, v []byte) error {
+			key := bytesutil.FromBytes8(k)
+			value, err := unmarshalSpan(ctx, v)
+			if err != nil {
+				return err
+			}
+			spanMap[key] = value
+			return nil
+		})
+	})
+	if spanMap == nil {
+		spanMap = make(map[uint64]types.Span)
+	}
+	return spanMap, err
+}
+
+// EpochSpanByValidatorIndex accepts validator index and epoch returns the corresponding spans
+// for slashing detection.
+// Returns error if the spans for this validator index and epoch does not exist.
+func (db *Store) EpochSpanByValidatorIndex(ctx context.Context, validatorIdx uint64, epoch uint64) (types.Span, error) {
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.EpochSpanByValidatorIndex")
+	defer span.End()
+	var err error
+	if db.spanCacheEnabled {
+		v, ok := db.spanCache.Get(epoch)
+		spanMap := make(map[uint64]types.Span)
+		if ok {
+			spanMap, ok = v.(map[uint64]types.Span)
+			if !ok {
+				return types.Span{}, cacheTypeMismatchError(v)
+			}
+			spans, ok := spanMap[validatorIdx]
+			if ok {
+				return spans, nil
+			}
+		}
+	}
+	var spans types.Span
+	err = db.view(func(tx *bolt.Tx) error {
+		b := tx.Bucket(validatorsMinMaxSpanBucket)
+		epochBucket := b.Bucket(bytesutil.Bytes8(epoch))
+		if epochBucket == nil {
+			return nil
+		}
+		key := bytesutil.Bytes8(validatorIdx)
```

Comment:
How large can one validators span be? In the future can we maybe try storing the full spans in one key and splitting them up, instead of having each epoch split up by a different key?

Reply:
validators span 7 bytes * 300K = 2.1 mb value
lets optimize it later

Reply:
wdyt?

Reply:
I think this is fine for now @0xKiwi

Reply:
Sounds good, definitely something we can look into later.